### PR TITLE
SIS images scraper

### PR DESCRIPTION
**Pull request by jn7070**
Scrapes images with RCS ID of each student in a course.

### CODE REVIEW

**Comment by jn7070** [comment]
Instructions on website: http://submitty.org/instructor/image_scraper

File: SIS_Images_Scraper.py [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+    driver.get("https://sis.rpi.edu/")
+
+    # if login is invalid with incorrect User ID or PIN, end the program
+    if not login(driver):
+        driver.close()
+
+    # if login is valid with correct User ID and PIN, continue the program by collecting data
+    else:
+        getInfoFromCourse(driver)
+        driver.close()
```
Comment (MasterOdin) [inline SIS_Images_Scraper.py]: You probably want to wrap a try/finally around this to always dispose of the driver, even if selenium crashes, or SIS dies in the middle of things or whatever.
```
try:
    driver.get('https://sis.rpi.edu/')
    if login(driver):
        getInfoFromCourse(driver)
finally:
    driver.quit()
```

**Review by holzbh — CHANGES_REQUESTED** [review]
>holzbh@[redacted]$ python SIS_Images_Scraper.py
>Traceback (most recent call last):
>  File "SIS_Images_Scraper.py", line 4, in <module>
>    from selenium.webdriver.support.ui import Select
>ImportError: No module named selenium.webdriver.support.ui
>holzbh@[redacted]:/mnt/d/SubmittyITools$ python3 SIS_Images_Scraper.py
>Traceback (most recent call last):
>  File "SIS_Images_Scraper.py", line 4, in <module>
>    from selenium.webdriver.support.ui import Select
>ImportError: No module named 'selenium'

It should be made clear which non-standard packages are pre-requisites. I'm also of the opinion the instructions/pre-reqs should make it clear if this is python2 or python3.

From WSL I'm probably not going to be able to run Chrome (or any browser really, though a text-based one like _links_ might work). [This](https://github.com/Microsoft/WSL/issues/1169) describes some of the challenges. If we expect users to use Chrome "Opening up SIS" isn't really a clear behavior - is this being done through Requests/sockets I don't see, or do we mean this is opening a browser and navigating to SIS? I'll have to toy around with this a little bit - probably get Python working in Windows?

I tried doing a `pip install --user chromedriver` and still get the same error:

>holzbh@[redacted]$ python3 SIS_Images_Scraper.py
>Traceback (most recent call last):
>  File "/home/holzbh/.local/lib/python3.5/site-packages/selenium/webdriver/common/service.py", line 76, in start
>    stdin=PIPE)
>  File "/usr/lib/python3.5/subprocess.py", line 947, in __init__
>    restore_signals, start_new_session)
>  File "/usr/lib/python3.5/subprocess.py", line 1551, in _execute_child
>    raise child_exception_type(errno_num, err_msg)
>FileNotFoundError: [Errno 2] No such file or directory: 'chromedriver'

>During handling of the above exception, another exception occurred:
>Traceback (most recent call last):
>  File "SIS_Images_Scraper.py", line 194, in <module>
>    driver = webdriver.Chrome()
>  File "/home/holzbh/.local/lib/python3.5/site-packages/selenium/webdriver/chrome/webdriver.py", >line 68, in __init__
>    self.service.start()
>  File "/home/holzbh/.local/lib/python3.5/site-packages/selenium/webdriver/common/service.py", line >83, in start
>    os.path.basename(self.path), self.start_error_message)
>selenium.common.exceptions.WebDriverException: Message: 'chromedriver' executable needs to be in >PATH. Please see https://sites.google.com/a/chromium.org/chromedriver/home

So I manually added it to the path, no good.

Next step was apt-get install chromium-chromedriver (which installs chromium-browser) as well.

After manually changing my $PATH post-install, I was able to get the script to run, though it just hangs.

I briefly considered getting WSL to play with a Windows chromedriver installation (you can kind of hack it - it won't like the .exe but you can make a file called `chromedriver` that just spawns the .exe, but at that point why bother with WSL?

So after all this, I went and installed Python 3.6.5 for Windows on my computer. And then it turned out I already had Python 3.6.2, it just wasn't in my root directory. So I stuck with 3.6.2.

I had to go look up where to get chromedriver, and then since a Windows user is even less likely to modify their PATH, I followed the suggestion [here](https://stackoverflow.com/questions/33150351/how-do-i-install-chromedriver-on-windows-10-and-run-selenium-tests-with-chrome) and just dropped it in my C:\windows directory (which already is on the PATH).

So finally, I got to a browser with SIS! It's not entirely clear to me that I should be watching the console, because I've got a Chrome browser that popped up on top and covered most of the console.

Normally I use a password manager, but not in Chrome, not sure how that'll mess with things if I have to launch my password manage, do authentication/2FA, and then get back to SIS. Might be worth testing.

Finally I played along, and got to the point where I said Yes to my first section of a course. A long pause later I got this:

>Traceback (most recent call last):
>  File "D:\SubmittyITools\SIS_Images_Scraper.py", line 203, in <module>
>    getInfoFromCourse(driver)
>  File "D:\SubmittyITools\SIS_Images_Scraper.py", line 187, in getInfoFromCourse
>    saveImagesToFolder(term, course, class_list)
>  File "D:\SubmittyITools\SIS_Images_Scraper.py", line 75, in saveImagesToFolder
>    os.makedirs(path, exist_ok=True)
>  File "C:\Users\[redacted]\AppData\Local\Programs\Python\Python36-32\lib\os.py", line 220, in makedirs
>    mkdir(name, mode)
>NotADirectoryError: [WinError 267] The directory name is invalid: 'Spring 2018/CSCI 1200 01: DATA >STRUCTURES, 51427 (30)'

That has so many characters Windows doesn't allow - you can see a full list [here](https://msdn.microsoft.com/en-us/library/windows/desktop/aa365247(v=vs.85).aspx#naming_conventions)

I think directory names should be shorter anyway, most of that information is extraneous.

====

Summary of issues:

1. Unclear what the pre-reqs are (Python 3, selenium), chromedriver installed and in PATH
2. Unclear that this will not work "headless" (requires a Chrome window) so you should not attempt this in WSL unless you want to do a lot of tweaks
3. In Windows at least, Chrome browser that spawns hides most of the console so it's hard to see the console wants input. The first time through I manually clicked Login and authenticated before thinking to check the console, so everything was in a weird state.
4. Script tries to make a directory with a very long name that has many illegal Windows characters.

**Comment by holzbh** [comment]
Just one more detail - you should use os.path.join() to avoid problems with "\\" vs "/".

**Comment by MasterOdin** [comment]
For points 2 & 3, you can add the `--headless` option:
```
from selenium import webdriver
from selenium.webdriver.chrome.options import Options
chrome_options = Options()
chrome_options.add_argument("--headless")
driver = webdriver.Chrome(chrome_options=chrome_options)
```

I agree with Buster to use a library to handle joining paths, but since this is just Python3.4+, use [pathlib](https://docs.python.org/3/library/pathlib.html)

--- 

However, given the comment on having to get chromedriver (or geckodriver if we wanted to use Firefox), however, it occurs to me that it'd actually make more sense to use [puppeteer](https://github.com/GoogleChrome/puppeteer) since installation of it (via `npm`) includes chromium for you and you wouldn't need to configure anything with your path or such, but this can be for a "second" version of it. We should just include instructions in the README on installation of requirements and such. Considering this, we probably want to put each "tool" into its own folder that has a README (how to run, install, etc.), installation script for dependencies, and anything else specific to that tool.

File: SIS_Images_Scraper.py [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+import urllib
+import os
+from selenium.webdriver.support.ui import Select
+from selenium import webdriver
+from selenium.webdriver.common.keys import Keys
+
+# Login to SIS
+def login(driver):
+    # Get User ID and PIN
+    user_id = getpass.getpass("User ID: ")
```
Comment (MasterOdin) [inline SIS_Images_Scraper.py]: Does this really need to be a getpass instance given how many other things we (at RPI) type our RIN into (and is probably stored as plaintext anyway). I've always hated how I couldn't see my RIN on this screen as well. Probably should just call in 'RIN' as well given how RPI specific SIS is.

File: SIS_Images_Scraper.py [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+        select_course = Select(driver.find_element_by_name('crn'))
+        options_course = select_course.options
+        course = options_course[index].text
+
+        # gets the images the user wants for the class section by looping until the user enters a valid command
+        foundAnswer = False
+        while not foundAnswer:
+            # Asks if user wants pictures from current course displayed
+            print("Do you want pictures from {}?".format(course))
+            answer = input("Y/N/Exit\n")
```
Comment (MasterOdin) [inline SIS_Images_Scraper.py]: `answer = input("Y/N/Exit\n").lower()` unless you're explicitly doing something with a chance of breaking someone's computer, please lowercase (or uppercase) an answer and then do the comparison.

**Review by MasterOdin — CHANGES_REQUESTED** [review]

**Comment by MasterOdin** [comment]
For renaming the course folders to something better (and doesn't have forbidden Windows symbols), we probably only really care about department code (4 characters), course code (four numbers), and section (probably only two numbers, but eh...), and so we can pull that info using the following regex `([A-Z]{4}) ([0-9]{4}) ([0-9]+)\:` and here's an example:
```
>>> a = re.match(r'([A-Z]{4}) ([0-9]{4}) ([0-9]+)\:', 'CSCI 1200 01: DATA >STRUCTURES, 51427 (30)')
>>> if a is None:
...     print("Invalid format for course name")
...
>>> a.groups()
('CSCI', '1200', '01')
>>> "{}-{}-{}".format(*a.groups())
'CSCI-1200-01'
```

Not sure how @holzbh  or @bmcutler care for having the semester folders sorted/renamed than just "Spring 2018" or "Fall 2018" or whatever.

**Comment by bmcutler** [comment]
I agree with making a folder for each tool and having a README.  But let's not duplicate the instructions.  In this case the README should just point to the submitty.org website instructions (& possibly also the raw markdown in the repo).

Usually this script will be run just for 1 semester at a time, the current semester, so no need to get fancy planning for multiple terms of data & sorting.  Most faculty have their own folder organization for terms & classes, so let's just leave it.

I agree with getting this version up for now, and version 2 can worry about porting to every OS/browser.

**Comment by holzbh** [comment]
Tested the newest commit, got the following results. I'm fine with this just working in WSL for the moment, but it's not quite there yet:

#### WSL, Spring 2018
First non-zero section it suggested (30 students) finished without errors. However, every PNG is a 68 byte file that definitely isn't the student's image. It looks like a rounded edge, possibly a layout image being grabbed? Attached one: [example](https://user-images.githubusercontent.com/26691043/40807986-cad19b08-64f3-11e8-9a2c-e29330e1b617.png)

#### WSL, Fall 2018
First non-zero section it suggested (70 students):

>Traceback (most recent call last):
>  File "SIS_Images_Scraper.py", line 221, in <module>
>    getInfoFromCourse(driver)
>  File "SIS_Images_Scraper.py", line 205, in getInfoFromCourse
>    class_list = getStudentInfoFromCourse(driver, select_course, index, class_list)
>  File "SIS_Images_Scraper.py", line 136, in getStudentInfoFromCourse
>    student.find_elements_by_tag_name('td')[1].find_element_by_class_name('fieldmediumtext').click()
>  File "/home/[redacted]/.local/lib/python3.5/site-packages/selenium/webdriver/remote/webelement.py", line 398, in find_element_by_class_name
>    return self.find_element(by=By.CLASS_NAME, value=name)
>  File "/home/[redacted]/.local/lib/python3.5/site-packages/selenium/webdriver/remote/webelement.py", line 653, in find_element
>    {"using": by, "value": value})['value']
>  File "/home/[redacted]/.local/lib/python3.5/site-packages/selenium/webdriver/remote/webelement.py", line 628, in _execute
>    return self._parent.execute(command, params)
>  File "/home/[redacted]/.local/lib/python3.5/site-packages/selenium/webdriver/remote/webdriver.py", line 314, in execute
>    self.error_handler.check_response(response)
>  File "/home/[redacted]/.local/lib/python3.5/site-packages/selenium/webdriver/remote/errorhandler.py", line 242, in check_response
>    raise exception_class(message, screen, stacktrace)
>selenium.common.exceptions.NoSuchElementException: Message: no such element: Unable to locate element: {"method":"class name","selector":"fieldmediumtext"}
>  (Session info: headless chrome=66.0.3359.181)
>  (Driver info: chromedriver=2.36 (0),platform=Linux 4.4.0-17134-Microsoft x86_64)

#### Windows, Spring 2018
First section it suggested (30 students)

>Traceback (most recent call last):
>  File "C:\Users\[redacted]\AppData\Local\Programs\Python\Python36-32\lib\urllib\request.py", line 1318, in do_open
>    encode_chunked=req.has_header('Transfer-encoding'))
>  File "C:\Users\[redacted]\AppData\Local\Programs\Python\Python36-32\lib\http\client.py", line 1239, in request
>    self._send_request(method, url, body, headers, encode_chunked)
>  File "C:\Users\[redacted]\AppData\Local\Programs\Python\Python36-32\lib\http\client.py", line 1285, in _send_request
>    self.endheaders(body, encode_chunked=encode_chunked)
>  File "C:\Users\[redacted]\AppData\Local\Programs\Python\Python36-32\lib\http\client.py", line 1234, in endheaders
>    self._send_output(message_body, encode_chunked=encode_chunked)
>  File "C:\Users\[redacted]\AppData\Local\Programs\Python\Python36-32\lib\http\client.py", line 1026, in _send_output
>    self.send(msg)
>  File "C:\Users\[redacted]\AppData\Local\Programs\Python\Python36-32\lib\http\client.py", line 964, in send
>    self.connect()
>  File "C:\Users\[redacted]\AppData\Local\Programs\Python\Python36-32\lib\http\client.py", line 1400, in connect
>    server_hostname=server_hostname)
>  File "C:\Users\[redacted]\AppData\Local\Programs\Python\Python36-32\lib\ssl.py", line 401, in wrap_socket
>    _context=self, _session=session)
>  File "C:\Users\[redacted]\AppData\Local\Programs\Python\Python36-32\lib\ssl.py", line 808, in __init__
>    self.do_handshake()
>  File "C:\Users\[redacted]\AppData\Local\Programs\Python\Python36-32\lib\ssl.py", line 1061, in do_handshake
>    self._sslobj.do_handshake()
>  File "C:\Users\[redacted]\AppData\Local\Programs\Python\Python36-32\lib\ssl.py", line 683, in do_handshake
>    self._sslobj.do_handshake()
>ssl.SSLError: [SSL: SSLV3_ALERT_HANDSHAKE_FAILURE] sslv3 alert handshake failure (_ssl.c:749)
>
>During handling of the above exception, another exception occurred:
>
>Traceback (most recent call last):
>  File "D:\SubmittyITools\SIS_Images_Scraper.py", line 221, in <module>
>    getInfoFromCourse(driver)
>  File "D:\SubmittyITools\SIS_Images_Scraper.py", line 209, in getInfoFromCourse
>    saveImagesToFolder(term, course, class_list)
>  File "D:\SubmittyITools\SIS_Images_Scraper.py", line 109, in saveImagesToFolder
>    urllib.request.urlretrieve(img_url, str(filepath))
>  File "C:\Users\[redacted]\AppData\Local\Programs\Python\Python36-32\lib\urllib\request.py", line 248, in urlretrieve
>    with contextlib.closing(urlopen(url, data)) as fp:
>  File "C:\Users\[redacted]\AppData\Local\Programs\Python\Python36-32\lib\urllib\request.py", line 223, in urlopen
>    return opener.open(url, data, timeout)
>  File "C:\Users\[redacted]\AppData\Local\Programs\Python\Python36-32\lib\urllib\request.py", line 526, in open
>    response = self._open(req, data)
>  File "C:\Users\[redacted]\AppData\Local\Programs\Python\Python36-32\lib\urllib\request.py", line 544, in _open
>    '_open', req)
>  File "C:\Users\[redacted]\AppData\Local\Programs\Python\Python36-32\lib\urllib\request.py", line 504, in _call_chain
>    result = func(*args)
>  File "C:\Users\[redacted]\AppData\Local\Programs\Python\Python36-32\lib\urllib\request.py", line 1361, in https_open
>    context=self._context, check_hostname=self._check_hostname)
>  File "C:\Users\[redacted]\AppData\Local\Programs\Python\Python36-32\lib\urllib\request.py", line 1320, in do_open
>    raise URLError(err)
>urllib.error.URLError: <urlopen error [SSL: SSLV3_ALERT_HANDSHAKE_FAILURE] sslv3 alert handshake failure (_ssl.c:749)>

**Pull request merged**